design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,235 @@
+# Solution Design - Inject Client Cert fix
+[//]: # (Change the title above from "Template" to your design's title) 
+
+## Table of Contents
+[//]: # (You can use this tool to generate a TOC - https://ecotrust-canada.github.io/markdown-toc/) 
+
+## Useful links
+[//]: # (Add links that may be useful for the reader)
+
+- [Existing PR with solution](https://github.com/cyberark/conjur/pull/1789)
+  - In this solution we wait for the timeout to get an answer and if we didn't
+    get one we create another WebSocket with an "ls" command to verify that the
+    cert is injected.
+- [Known issue in k8s exec](https://github.com/kubernetes/kubernetes/issues/89899)
+
+## Background
+[//]: # (Give relevant background for the designed feature. What is the motivation for this solution?)
+We have an issue with the "inject client cert" process where we don't always return
+a correct response to the client regarding the injection success. The injection is
+done via a WebSocket to Kubectl Exec where we use a "tar" command and the cert content in STDIN.
+There is [an issue](https://github.com/kubernetes/kubernetes/issues/89899) in k8s
+where messages that are sent via WebSockets to Kubectl Exec are staying hanged without
+returning a response when using STDIN. 
+
+## Issue description
+[//]: # (Elaborate on the issue you are writing a solution for)
+The current behaviour is that we close the WebSocket connection after we send
+the injection request to kubectl. This is not correct as this
+way we do not wait for kubectl to send its response.
+
+This has 2 implications:
+1. If the injection request has succeeded we will return a 200 OK response
+   to the 'inject_client_cert' request even if the cert has not finished to
+   be written to the pod. This can lead to a race-condition where the client
+   will try to read the cert before it is there.
+1. If the injection request has failed in the k8s API, we will not
+   write the error to the Conjur logs, and we will falsely return a 200
+   OK response to the 'inject_client_cert' request
+
+We would like to find a solution that will indicate whether the cert was successfully
+injected and log the error otherwise.
+
+## Solution
+[//]: # (Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breakdown of the feature, but just a recommendation to how the solution should be approached.)
+
+We would like to inject the cert and get an indication that the cert was injected 
+or the error if it was not. Let's explore the different options that we have to inject the cert. 
```

Comment:
@orenbm Did you consider replacing the websocket solution with POST HTTP request?
In the same way `kubectl` does it.
If it's good for `kubectl` it should be good for us, no?
It uses SDPY protocol, not WebSocket.
Read here for more info: https://erkanerol.github.io/post/how-kubectl-exec-works/

Reply:
Thanks @abrahamko !

It looks like the SPDY protocol is being [deprecated](https://github.com/kubernetes/enhancements/issues/384) so it's best not to use it. As written [here](https://www.openshift.com/blog/executing-commands-in-pods-using-k8s-api):
```
kubectl and oc use the SPDY protocol at the moment, which is being deprecated . The second option is to use Websockets, which seems to be the best way. Anyway one of these two protocols, SPDY or WebSockets, is required for communication with this endpoint, and the API will refuse requests without Upgrade headers.
```